documentation.md

```diff
@@ -0,0 +1,205 @@
+
+The specification is developed by EMVCo, best described by [their own
+words](https://www.emvco.com/about/overview/):
+
+> EMVCo exists to facilitate worldwide interoperability and acceptance of
+> secure payment transactions
+
+The specification is public and can be found at https://www.emvco.com/emv-technologies/3d-secure/.
+This service supports all active versions of 3-D Secure version 2:
+
+- 2.1.0
+- 2.2.0
+
+During integration, you will need to refer to specification for _both_
+versions.  The EMVCo 3-D Secure has a different program name for each card
+scheme, this service supports the following card schemes:
+
+
+| Card Scheme     | Program name                                                 |
+| --------------- | ------------------------------------------------------------ |
+| Mastercard      | Mastercard Identity Check (previously Mastercard SecureCode) |
+| Visa            | Visa Secure (previously Verified by Visa)                    |
+
+<!---
+}}}
+-->
+
+## Getting Started
+
+### What is an authentication?
+
+An authentication in 3-D Secure is the process of verifying cardholder
+involvement in e.g. a purchase. An authentication results from and
+_authentication flow_ and proof of authentication is an
+_authentication value_.
+
+This figure illustrates the simplified authentication flow from a requestor
+point of view:
+
+<div class="img">
+  <img src="authentication.svg" alt="Authentication"/>
+</div>
+
+An _authentication flow_ is one of the following categories:
+* _Frictionless flow_
+
+   This flow includes messages 1-4, where the `authenticationValue`, `eci` and
+   `dsTransID` are returned in 4.
+
+* _Challenge flow_
+
+   This includes 1-8, where the `authenticationValue`, `eci` and `dsTransID`
+   are returned in 8.
+
+* _Out-of-band (OOB)_
+
+   This includes 1-8, where the `authenticationValue`, `eci` and `dsTransID`
+   are returned in 8.
+
+* _Decoupled Authentication_
+
+   This includes 1-4,6-8, where the `authenticationValue`, `eci` and `dsTransID`
+   are returned in 8.
+
+### Performing an authentication flow
+
+1. Use the [`/preauth`](#the-preauth-call) call to determine 3-D Secure version.
+2. Perform an authentication using the [`/auth`](#the-auth-call) call.
+3. If `transStatus` is `C`, perform a [challenge
+   flow](#performing-the-challenge) on the cardholder device.
+4. Retrieve values from [`/postauth`](#the-postauth-call) endpoint after
+   challenge flow has completed.
+
+## Examples
+
+### Frictionless flow example
+
+**Card Verification:**
+```
+$ APIKEY=********-****-****-****-************
+$ PAN=****************
+$ curl -H 'APIKey: $APIKEY' -d "{\"acctNumber\": \"$PAN\"}" https://service.staging.3dsecure.io/preauth
+```
+
+## Explanation
+
+TODO: Section is currently, envisioned is a sort of FAQ.
+
+## Reference
+
+### Authentication
+
+Access to the service is granted by an API key. The API key is used on each request using the
+`APIKey` HTTP header as such:
+```
+APIKey: <YOUR UUID API KEY>
```

Comment:
The answer is both syntax and semantics.

1. There is no password, there is only the API key. Which is not the same as a user name as a user can have several API keys.
2. I remember several times having to explain to people that they need  to remember the `:` when integrating against the gateway. Though many libraries likely support some HTTP Basic Auth.
3. I don't want someone to perform the call from the browser, e.g. attempt to open the URL and add the API Keys themselves.

But I'm not married to an `APIKey` header.

I found out yesterday that it also matches e.g. what Adyen [does](https://docs.adyen.com/api-explorer/#/PaymentSetupAndVerificationService/v50/overview).

Reply:
This is an entrance to the JWT consideration (where other systems create long lived JWTs, e.g. with expiry until next payment date).

* Re. 1. and 2. If we have a user concept, then, do we have a username? If so, Basic Auth with `username:apikey` could make sense?

* Re. 3. Why? One could consider it a feature which could be handy for testing — this is basically what's utilized via `https://mpi.3dsecure.io/secure3d_auth`.

Reply:
* 1. and 2. It could, but we would require the username for no reason, as it is only used in the dashboard. Also: That would mean to semantically do a user lookup by password, which is not the case.
* 3. We can consider creating an endpoint like this, however it is not as simple in 3DSv2. Here we need 2 callback URLs and a lot of data to be entered.

Reply:
* Re. 1 and 2: Good point! Also, the API key is not really a password. However, we could use the username as the thing that refers to the user, and then in fact use a password (which we then only store e.g. `scrypt`'ed). However, I like the JWT strategy much better.
* Re. 3: We can just populate it with dummy data (and allow to store data in a cookie). It is quite essential that we can do example flows easily; it'll also be a reference implementation for clients to be inspired by.